### PR TITLE
[deckhouse] one source of truth for deckhouse registry (stage #2)

### PR DESCRIPTION
**Pull request by fl64**
## Description
Removed the `deckhouse-registry` secret mount from the deckhouse deployment. 
Added validating deckhouse image in `change-registry.sh` script.

## Why do we need it, and what problem does it solve?
Closes https://github.com/deckhouse/deckhouse/issues/714

## What is the expected result?
The `deckhouse-registry` secret is no longer mounted in the `deckhouse` deployment. You must now specify dockerconfig when using `GetRegistryClient`.
`change-registry.sh` script checks for a deckhouse image with the current tag in the new registry before switching to new registry.

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse
type: feature
summary: Added validating deckhouse image in `change-registry.sh` script.
impact_level: default 
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @fl64. I've started the workflow for label 'e2e/run/openstack'...

File: modules/002-deckhouse/hooks/set_module_image_value.go [hunk 1]
```diff
@@ -69,9 +71,18 @@ func parseDeckhouseImage(input *go_hook.HookInput) error {
 	}
 	image := deckhouseSnapshot[0].(string)
 
+	parts := strings.SplitN(image, ":", 2)
```
Comment (name212) [inline modules/002-deckhouse/hooks/set_module_image_value.go:74]: Deckhouse image can contain port. for example `air-gap-domain.com:7777/deckhouse/fe:v.1.42.0` It is important for air gap environments. We should support it. This code split string for only 2 parts.

File: modules/002-deckhouse/hooks/set_module_image_value_test.go [hunk 2]
```diff
@@ -62,7 +62,7 @@ spec:
 				Expect(f).To(ExecuteSuccessfully())
 				deployment := f.KubernetesResource("Deployment", "d8-system", "deckhouse")
 				Expect(deployment.Exists()).To(BeTrue())
-				Expect(f.ValuesGet("deckhouse.internal.currentReleaseImageName").String()).To(Equal("registry.deckhouse.io/deckhouse/ce:test"))
+				Expect(f.ValuesGet("deckhouse.internal.currentReleaseImageName").String()).To(Equal("registry.deckhouse.io/deckhouse/fe:test"))
```
Comment (name212) [inline modules/002-deckhouse/hooks/set_module_image_value_test.go:65]: Add test for 'with port' case

File: tools/change-registry.sh [hunk 3]
```diff
@@ -85,6 +85,7 @@ function parse_args() {
   REGISTRY_ADDRESS="$(cut -d "/" -f1 <<< "$TEMP_URL")"
   REGISTRY_PATH="${TEMP_URL#"$REGISTRY_ADDRESS"}"
   REGISTRY_SCHEME="$(sed "s/:\/\///" <<< "$REGISTRY_SCHEME")"
+  DECKHOUSE_TAG="$(kubectl -n d8-system get deploy deckhouse -o json | jq '.spec.template.spec.containers[] | select(.name=="deckhouse") | .image | split(":")[1]' -r)"
```
Comment (name212) [inline tools/change-registry.sh:88]: Need support image paths with port

File: tools/change-registry.sh [hunk 4]
```diff
@@ -118,6 +119,16 @@ function parse_args() {
     exit 1
   fi
```
Comment (name212) [inline tools/change-registry.sh:120]: Also please fix [this](https://github.com/deckhouse/deckhouse/pull/3499/files#diff-8bf4756771579a6415a1f1eef7cc430258b62a822e0470732b116b6c4c45d631R105) validator for support images paths with port

**Review by name212 — CHANGES_REQUESTED** [review]
- remove unused constant
- support image path with port

File: tools/change-registry.sh (hunk 4 above)
Comment (name212) [reply tools/change-registry.sh:120]: https://github.com/deckhouse/deckhouse/blob/8efbfc6b68208d461c839379c62ffcc7d1f965f9/tools/change-registry.sh#L105

Comment (name212) [reply tools/change-registry.sh:120]: And one thing. Do you check changing registry operation with CA file?

File: dhctl/pkg/kubernetes/actions/manifests/manifests.go [hunk 5]
```diff
@@ -118,33 +118,6 @@ func ParameterizeDeckhouseDeployment(input *appsv1.Deployment, params DeckhouseD
 		deckhousePodTemplate.Spec.ImagePullSecrets = []apiv1.LocalObjectReference{
 			{Name: "deckhouse-registry"},
 		}
-
```
[the remark this answers is not in the document]
Comment (fl64) [reply dhctl/pkg/kubernetes/actions/manifests/manifests.go:121]: fixed

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @fl64. I've started the workflow for label 'e2e/run/openstack'...

**Review by name212 — DISMISSED** [review]

File: go_lib/dependency/cr/cr.go [hunk 6]
```diff
@@ -112,12 +112,12 @@ func (r *client) Digest(tag string) (string, error) {
...
 
-func readAuthConfig(configPath string) (authn.AuthConfig, error) {
-	dockerConfigBytes, err := os.ReadFile(configPath)
+func readAuthConfig(dockerCfgBase64 string) (authn.AuthConfig, error) {
+	dockerCfg, err := base64.StdEncoding.DecodeString(dockerCfgBase64)
 	if err != nil {
 		return authn.AuthConfig{}, err
 	}
-	auths := gjson.GetBytes(dockerConfigBytes, "auths").Map()
+	auths := gjson.Get(string(dockerCfg), "auths").Map()
```
Comment (yalosev) [inline go_lib/dependency/cr/cr.go:120]: handle this situation
```
{
  "auths": {
    "dev-registry.deckhouse.io": {
      "auth": "d3JpdGV0b2tlbjo0YWM5ZmUyYjcwYzU0ZDA5ZjZkNDM1MzI4ZGZmMDg0NA=="
    },
     "prod-registry.deckhouse.io": {
      "auth": "d3JpdGV0b2tlbjo0YWM5ZmUyYjcwYzU0ZDA5ZjZkNDM1MzI4ZGZmMDg0NA=="
    }
  }
}
```

File: go_lib/dependency/cr/cr.go [hunk 7]
```diff
@@ -112,12 +112,12 @@ func (r *client) Digest(tag string) (string, error) {
 	return d.String(), nil
 }
 
-func readAuthConfig(configPath string) (authn.AuthConfig, error) {
-	dockerConfigBytes, err := os.ReadFile(configPath)
+func readAuthConfig(dockerCfgBase64 string) (authn.AuthConfig, error) {
+	dockerCfg, err := base64.StdEncoding.DecodeString(dockerCfgBase64)
```
Comment (yalosev) [inline go_lib/dependency/cr/cr.go:116]:
````
u1, _ := url.Parse(repo)
u1.Host
```

key := u1.Host

  var tmp map[string]interface{}
  json.Unmarshal(a, &tmp)

for _, a := range tmp {
  u2, _ := url.Parse(a)
  if u2 != u1 { 
    continue
  }
...
}

File: go_lib/dependency/cr/cr.go [hunk 8]
```diff
@@ -48,7 +48,7 @@ type client struct {
 
 // NewClient creates container registry client using `repo` as prefix for tags passed to methods. If insecure flag is set to true, then no cert validation is performed.
 // Repo example: "cr.example.com/ns/app"
-func NewClient(repo string, options ...Option) (Client, error) {
+func NewClient(repo, dockerCfgBase64 string, options ...Option) (Client, error) {
```
Comment (yalosev) [inline go_lib/dependency/cr/cr.go:51]:
```
func WithAuth(dockerCfg string) Option {
	return func(options *registryOptions) {
		options.dockerCfg = dockerCfg
	}
}
```

repo may be `registry.deckhouse.io/ee/modules`

**Review by yalosev — CHANGES_REQUESTED** [review]

**Review by yalosev — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/yandex-cloud'...

File: modules/002-deckhouse/hooks/set_module_image_value_test.go [hunk 9]
```diff
@@ -62,7 +62,38 @@ spec:
...
+        image: registry.deckhouse.io:666/deckhouse/ce:test666
+`, 1))
+				f.RunHook()
+			})
+
+			It("Should run", func() {
+				Expect(f).To(ExecuteSuccessfully())
+				deployment := f.KubernetesResource("Deployment", "d8-system", "deckhouse")
+				Expect(deployment.Exists()).To(BeTrue())
+				Expect(f.ValuesGet("deckhouse.internal.currentReleaseImageName").String()).To(Equal("registry.deckhouse.io/deckhouse/fe:test666"))
```
Comment (name212) [inline modules/002-deckhouse/hooks/set_module_image_value_test.go:96]: it should be equal `registry.deckhouse.io:666/deckhouse/ce:test666` not `registry.deckhouse.io/deckhouse/fe:test666`. Or does not equal? and why?

**Review by name212 — APPROVED** [review]

**Pull request merged**